### PR TITLE
feat(source): support source failover & refactor

### PR DESCRIPTION
**Pull request by tabVersion**
Signed-off-by: tabVersion <tabvision@bupt.icu>

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

support failover. 

* introduce `DummyConnector`, it will serve as an idle connector for the executor which is assigned no split.

---

rename SourceSplit -> SplitMetaData:

```
- fn to_string(&self) -> Result<String>
+ fn to_json_bytes(&self) -> Result<Bytes>
```

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)

#1963

### CODE REVIEW

File: src/connector/src/state.rs [hunk 1]
```diff
@@ -183,20 +187,19 @@ mod tests {
...
+    impl SplitMetaData for TestSourceState {
+        fn id(&self) -> String {
             self.partition.clone()
         }
 
-        fn encode(&self) -> Bytes {
-            let bytes = self.offset.to_le_bytes();
-            Bytes::copy_from_slice(bytes.clone().as_ref())
-            // Bytes::from(self.offset.clone().to_string())
+        fn to_json_bytes(&self) -> Result<Bytes> {
```
Comment (neverchanje) [inline src/connector/src/state.rs:195]:
```suggestion
        fn to_json_bytes(&self) -> Bytes {
```

Impossible to fail?

File: src/connector/src/base.rs [hunk 2]
```diff
@@ -47,20 +47,36 @@ pub struct SourceMessage {
...
 /// The persistent state of the connector.
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct ConnectorState {
     pub identifier: Bytes,
     pub start_offset: String,
     pub end_offset: String,
 }
 
+impl SplitMetaData for ConnectorState {
```
Comment (neverchanje) [inline src/connector/src/base.rs:64]: I think ConnectorState is not a SplitMetaData? It's not going to be used by `list_splits()`.

```suggestion
impl ConnectorState {
```

Comment (tabVersion) [reply src/connector/src/base.rs:64]: The contents of the `SourceStateHandler` store need to implement the trait.

File: src/connector/src/state.rs (hunk 1 above)
Comment (tabVersion) [reply src/connector/src/state.rs:195]: Looks true, will fix

File: src/connector/src/base.rs [hunk 3]
```diff
@@ -47,22 +52,93 @@ pub struct SourceMessage {
...
+        }
+        let mut connector_states: Vec<Self> = Vec::with_capacity(state.len());
+        for (split, offset) in state {
+            connector_states.push(Self {
+                identifier: Bytes::from(split),
+                start_offset: offset.clone(),
+                end_offset: "".to_string(),
+            })
+        }
+        connector_states
```
Comment (neverchanje) [inline src/connector/src/base.rs:82]:
```suggestion
        let mut connector_states: Vec<Self> = Vec::with_capacity(state.len());
        connector_states.extend(state.iter().map(|(split, offset)| {
            Self {
                identifier: Bytes::from(split),
                start_offset: offset.clone(),
                end_offset: "".to_string(),
            }
        }))
        connector_states
```

Use extend whenever it's possible.

File: src/connector/src/base.rs [hunk 4]
```diff
@@ -47,22 +52,93 @@ pub struct SourceMessage {
...
     pub start_offset: String,
     pub end_offset: String,
 }
 
+impl ConnectorState {
+    pub fn from_hashmap(state: HashMap<String, String>) -> Vec<Self> {
+        if state.is_empty() {
+            return vec![];
+        }
+        let mut connector_states: Vec<Self> = Vec::with_capacity(state.len());
```
Comment (neverchanje) [inline src/connector/src/base.rs:74]: So actually ConnectorState means the state of one single split? Then why not call it SplitState? It makes me confuse to think of ConnectorState is the state of an entire connector.

Comment (neverchanje) [reply src/connector/src/base.rs:74]: Also, SourceStateHandler should better be named SplitStateHandler or SplitStateStore

Comment (tabVersion) [reply src/connector/src/base.rs:77]: let's unite the state-related stuff in another pr.

I plan to remove `ConnectorState` and use the connector-specific split description for state management.

**Review by neverchanje — APPROVED** [review]
lgtm, `StreamChunkWithState` looks weird to me but anyway we can probably try another approach later.

**Pull request merged**